File: desktop - Linux, Mac, Windows/Internet.py
import requests
import socket
import time
from notification import show_notification
from ErrorType import ErrorType
from logger import logger

from PyQt5.QtCore import QObject, pyqtSlot

class Internet(QObject):
    def __init__(self):
        QObject.__init__(self)
        self.isInternet_monitor_stopped = False

    # websites request
    websites = {    # replace values with websites of which you would like to send response
         0:  'www.kamluk.wd.nubip.edu.ua',
         1:  'www.google.com',
         2:  'www.duckduckgo.com'
    }

    # -----
    logger.debug('Provided URLs to connection >\n %s', websites)
    # -----

    def disable_monitoring(self, isInternet_monitor_stopped):
        self.isInternet_monitor_stopped = isInternet_monitor_stopped

    def get_by_socket_connection_website(self, website_url):
        try:
            host = socket.gethostbyname(website_url)
            internet_socket = socket.create_connection((host, 80),2)
            internet_socket.close()
            return True
        except:
            pass
        return False

    def check_internet_connection(self):
        # before start we need to check if we have internet
        internet_current_status = self.get_by_socket_connection_website(self.websites[0])
        # -----
        logger.warning('At launch have an Internet connection > %s', internet_current_status)
        # -----
        while not self.isInternet_monitor_stopped:
            for i in range (len(self.websites)):
                while not self.isInternet_monitor_stopped:
                     website_conn_established = self.get_by_socket_connection_website(self.websites[i])

                     if internet_current_status != website_conn_established:
                         internet_current_status = website_conn_established
                         show_notification(ErrorType(website_conn_established).error_title(), self.websites[i])

                     if not website_conn_established:
                         break

                     time.sleep(2)  # Delay 2s to show notification

File: desktop - Linux, Mac, Windows - C++/internet.cpp
#include "internet.h"

#include <QThread>
#include <QTcpSocket>
#include <QDebug>


Internet::Internet(QWidget *parent) : QWidget(parent)
{
    isInternet_monitor_stopped = false;
}

void Internet::disable_monitoring(bool stop_monitoring)
{
    this->isInternet_monitor_stopped = stop_monitoring;
}

InternetStatus Internet::get_by_socket_connection_website(std::string website_url)
{
      qDebug() <<"Inside socket function";
      QTcpSocket *socket = new QTcpSocket;
      socket->connectToHost(QString(website_url.c_str()), 80);
      if (socket->waitForConnected(3000))
      {
        socket->close();
        return INTERNET_CONNECTED;
      }
     delete socket;
     return INTERNET_NOT_CONNECTED;
}

void Internet::check_internet_connection()
{
   // before start we need to check if we have internet
   InternetStatus internet_current_status = get_internet_connection_at_launch(WEBSITES[0]);

    while (!isInternet_monitor_stopped)
    {
        for (unsigned short i=0; i<WEBSITES.size(); i++)
        {
            while(!isInternet_monitor_stopped)
            {
                InternetStatus website_conn_established = get_by_socket_connection_website(WEBSITES[i]);

                if (internet_current_status != website_conn_established)
                {
                    internet_current_status = website_conn_established;
                    emit internet_status_changed(internet_current_status,
                                                 QString::fromUtf8((WEBSITES[i].c_str())));
                    qDebug()<< "Internet Connection"<<internet_current_status;
                }

                if (!website_conn_established)
                    break;

                // delay 3s to stop monitoring while notification is showing
                QThread::sleep(3);
            }
        }       
    }

    if (isInternet_monitor_stopped)
         emit internet_stopped_monitored(INTERNET_NOT_MONITORED);
}

InternetStatus Internet::get_internet_connection_at_launch(std::string website_url)
{
     InternetStatus internet_status = get_by_socket_connection_website(website_url);
     emit internet_status_at_launch(internet_status);
     qDebug() << "At launch have an Internet connection " << internet_status;
     return internet_status;
}



